data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java

```diff
@@ -100,64 +102,95 @@ public void initialise(final File inputFile) throws IOException {
+        signingHistory.signedBlocks.stream()
+            .map(SignedBlock::getSlot)
+            .filter(Objects::nonNull)
+            .max(UInt64::compareTo);
+    final Optional<UInt64> sourceEpoch =
+        signingHistory.signedAttestations.stream()
+            .map(SignedAttestation::getSourceEpoch)
+            .filter(Objects::nonNull)
+            .max(UInt64::compareTo);
+    final Optional<UInt64> targetEpoch =
+        signingHistory.signedAttestations.stream()
+            .map(SignedAttestation::getTargetEpoch)
+            .filter(Objects::nonNull)
+            .max(UInt64::compareTo);
+    final ValidatorSigningRecord record =
+        new ValidatorSigningRecord(
+            metadata.genesisValidatorsRoot,
+            lastSlot.orElse(UInt64.ZERO),
+            sourceEpoch.orElse(ValidatorSigningRecord.NEVER_SIGNED),
+            targetEpoch.orElse(ValidatorSigningRecord.NEVER_SIGNED));
+    return new SigningHistory(signingHistory.pubkey, record);
+  }
+
+  /**
+   * Update local slashing protection data with everything from the source data given to the
+   * initialise function.
+   *
+   * @param statusConsumer Consumer of any status strings that are generated
+   * @return Any errors for specific keys will be returned in a map, otherwise an empty map if there
+   *     are no errors.
+   */
+  public Map<BLSPublicKey, String> updateLocalRecords(final Consumer<String> statusConsumer) {
+    final Map<BLSPublicKey, String> errors = new HashMap<>();
+    data.forEach(
+        record -> {
+          Optional<String> error = updateLocalRecord(record, statusConsumer);
+          error.ifPresent(errorString -> errors.put(record.pubkey.asBLSPublicKey(), errorString));
+        });
+    statusConsumer.accept("Updated " + data.size() + " validator slashing protection records");
+    if (errors.size() > 0) {
+      statusConsumer.accept("There were " + errors.size() + " errors found during import.");
     }
+    return errors;
   }
 
-  public void updateLocalRecords(final Path slashingProtectionPath) {
-    this.slashingProtectionPath = slashingProtectionPath;
-    data.forEach(this::updateLocalRecord);
-    log.display("Updated " + data.size() + " validator slashing protection records");
+  /**
+   * Update local slashing protection data for a specific public key
+   *
+   * @param publicKey The public key to load if present in the source data given to the initialise
+   *     function.
+   * @param statusConsumer Consumer of any status strings that are generated
+   * @return Any error will be returned, otherwise an empty response on successful load.
+   */
+  public Optional<String> updateSigningRecord(
+      final BLSPublicKey publicKey, final Consumer<String> statusConsumer) {
+    final BLSPubKey key = new BLSPubKey(publicKey);
+    return data.stream()
+        .filter(signingHistory -> signingHistory.pubkey.equals(key))
+        .findFirst()
+        .flatMap(record -> updateLocalRecord(record, statusConsumer));
```

Comment:
Is it guaranteed that there's only one record per validator in a slashing protection file? It's usual to have it appear multiple times but I'd have thought it was valid.

So it think we'd need to swap these around:
```suggestion
        .flatMap(record -> updateLocalRecord(record, statusConsumer))
        .findFirst();
```

That way we import every record or stop the first time one returns an error and return that error.

Reply:
A test for having two records for the same validator would be a good idea too I suspect.